app/grandchallenge/participants/forms.py

```diff
@@ -1,8 +1,86 @@
+from django.core.exceptions import ValidationError
+from django.db.transaction import on_commit
 from django.forms import HiddenInput, ModelForm, TextInput
 
 from grandchallenge.core.forms import SaveFormInitMixin
 from grandchallenge.core.widgets import JSONEditorWidget
-from grandchallenge.participants.models import RegistrationQuestion
+from grandchallenge.participants.form_fields import RegistrationQuestionField
+from grandchallenge.participants.models import (
+    RegistrationQuestion,
+    RegistrationQuestionAnswer,
+    RegistrationRequest,
+)
+
+
+class RegistrationRequestForm(ModelForm):
+    class Meta:
+        model = RegistrationRequest
+        fields = (
+            "user",
+            "challenge",
+        )
+        widgets = {
+            "user": HiddenInput(),
+            "challenge": HiddenInput(),
+        }
+
+    def __init__(self, *args, challenge, user, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.fields["user"].initial = user
+        self.fields["user"].disabled = True
+        self.fields["challenge"].initial = challenge
+        self.fields["challenge"].disabled = True
+
+        self._registration_questions = RegistrationQuestion.objects.filter(
+            challenge=challenge
+        )
+
+        for q in self._registration_questions:
+            self.fields[str(q.pk)] = RegistrationQuestionField(
+                registration_question=q
+            )
+
+    @property
+    def _registration_question_answers(self):
+        for q in self._registration_questions:
+            yield RegistrationQuestionAnswer(
+                question=q,
+                registration_request=self.instance,
+                answer=self.cleaned_data.get(str(q.pk), ""),
+            )
+
+    def clean(self):
+        result = super().clean()
+        for answer in self._registration_question_answers:
+            try:
+                answer.full_clean(
+                    exclude=[
+                        "registration_request",  # Not saved at this point yet
+                    ],
+                )
+            except ValidationError as e:
+                answer_error = e.error_dict.get("answer")
+                if answer_error:
+                    self.add_error(str(answer.question.pk), answer_error)
+                else:
+                    raise e
+        return result
+
+    def _save_questions(self):
+        self.instance.refresh_from_db()
+        try:
+            for answer in self._registration_question_answers:
+                answer.full_clean()
+                answer.save()
+        except Exception as e:
+            self.instance.delete()  # Also deletes already created answers
+            raise e
+
+    def save(self, *args, **kwargs):
+        result = super().save(self, *args, **kwargs)
+        on_commit(self._save_questions)
+        return result
```

Comment:
This construction is a bit awkward, you are using one form to create multiple objects. Unsure on the solution, but I think Django has something built in for this: [formsets](https://docs.djangoproject.com/en/5.1/topics/forms/formsets/). You would then have one form per answer which would be individually validated.

Reply:
It is awkward; I get the sense that generally the practice is to have a Form per object/model. I'll see if I can introduce formsets to make that happen. It did take a non-trivial amount of tinkering to get current form to behave as it should, we'll see.

Reply:
Yeah, I'm not sure if it is the solution. We only have them in one other place in the codebase, and there you are adding multiple of the same objects to one other, here the number would be fixed. Good to research this.

Reply:
See `QuestionForm` and the `option_form_set`. Happy to walk you through that.

Reply:
From what I understand, the main objection was that a single `Form` created multiple objects. 

The way that the `QuestionForm` goes about it is that it is still within the `Form`, creating multiple objects. The only difference with the current construct in this PR is that you'd put a `FormSet -> Form` as an extra layer. 

I am currently exploring the option of having the `View` hold both a form and a modelformset, which comes with slightly different set 'tee' functions where you call things both the form and the modelformset.

Thoughts?

Reply:
I think you misunderstand the concept.

Now you are validating and saving Answer objects using a Question form.

Formsets move validation of individual objects to their own forms. So, you would have a formset of answer forms, each answer form is responsible for validating and saving its answer.

You still need to hookup the validation and saving of the formset from the parent form - this is where the link to the registration request comes in of course - but it means the validation and saving of answers is pushed down to an answer form which is where it should be.

Reply:
Views shouldn't mess about with forms, views should be extremely simple - does the user have permission? Then hook them up with a form and template. Redirect them.

Reply:
Ah. So it's more about having a dedicated `Form` perform the saving per object. Sure, adding some boilerplate that pushes the field and validation into a Form works, and using a formal formset is reasonably trivial.

Reply:
I refactored the form to use the fromsets. Using them is a reasonable fit. Had to make two 'fixes':
- Force the use of the required attribute, this only plays a part for formats when you don't want to create objects for empty forms. Having a 'required' attribute in the input would prevent posting empty forms.
- Force saving of empty forms, empty answers should be explicit and not implicit by missing answers.